### PR TITLE
rohd-hcl: find 

### PR DESCRIPTION
**Pull request by RPG-coder-intc**
## Description & Motivation

Find
- Count 0s or 1s
- Find First 0s or 1s
- Find Nth 0s or 1s

## Testing

Added test under `find_test.dart`

## Documentation

>TBD

### CODE REVIEW

**Comment by RPG-coder-intc** [comment]
@mkorbel1 here is my PR :)

File: lib/src/find.dart [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+// fifo.dart
+// Implementation of FIFOs.
+//
+// 2023 March 13
+// Author: Rahul Gautham Putcha <rahul.gautham.putcha@intel.com>
+//
+
+// import 'dart:collection';
+
+// ignore_for_file: lines_longer_than_80_chars
```
Comment (RPG-coder-intc) [inline lib/src/find.dart:13]: remove

File: lib/src/find.dart [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    _output <=
+        (countOne
+            // count one
+            ? count
+            // Count zero by removing one's from bus width
+            : Const(bus.width, width: count.width) - count);
+  }
+}
+
+/*
```
Comment (RPG-coder-intc) [inline lib/src/find.dart:160]: remove

File: lib/src/find.dart [hunk 3]
```diff
@@ -0,0 +1,154 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// fifo.dart
```
Comment (RPG-coder-intc) [inline lib/src/find.dart:4]: find

File: lib/src/find.dart [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+// Implementation of FIFOs.
```
Comment (RPG-coder-intc) [inline lib/src/find.dart:5]: Find functionality

File: lib/src/find.dart [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+//
+// 2023 March 13
```
Comment (RPG-coder-intc) [inline lib/src/find.dart:7]: today (2023 July 11)

File: lib/src/find.dart [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    final bin = OneHotToBinary(oneHotList.rswizzle()).binary;
+    _output = addOutput('findNthOne', width: bin.width);
+    _output <= bin;
+  }
+}
+
+/// Doc coming soon...
+class Count extends Module {
```
Comment (mkorbel1) [inline lib/src/find.dart:131]: Maybe we should put count stuff in a different file?

File: lib/src/find.dart [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+/// Doc coming soon...
+class FindFirstOne extends Module {
+  /// Doc coming soon...
+  late Logic _output;
+
+  /// Doc coming soon...
+  Logic get index => _output;
+
+  /// Doc coming soon...  First 1 from the least significan (right)
+  FindFirstOne(Logic bus) {
```
Comment (mkorbel1) [inline lib/src/find.dart:26]: Perhaps all these `Find`s can be collapsed into one class?  Finding the N'th of something can be used to find the 0'th of it, and finding 0's and 1's is pretty similar.

File: lib/src/find.dart [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  ///
+  /// By default [Find] will look for first search parameter `1` or `0`.
+  /// If [n] is given, [Find] an [n]th search from first
+  /// occurance
+  Find(Logic bus, {bool one = true, Logic? n}) {
+    if (n != null) {
+      final oneHotList = <Logic>[];
+      for (var i = 0; i < bus.width; i++) {
+        if (one && i == 0) {
+          oneHotList.add(~bus[i] & n.eq(0));
```
Comment (mkorbel1) [inline lib/src/find.dart:38]: what if `~one && i==0`?  seems like missing test case?

File: lib/src/find.dart [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        } else {
+          final zeroCount = Count(bus.getRange(0, i + 1), countOne: one);
```
Comment (mkorbel1) [inline lib/src/find.dart:40]: let's name the variable something more flexible like just `count` or something, since this could be counting 1s or 0s

File: lib/src/find.dart [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+          oneHotList.add(busCheck);
+        } else {
+          final rangeCheck =
+              one ? ~bus.getRange(0, i).or() : bus.getRange(0, i).and();
+          oneHotList.add(busCheck & rangeCheck);
+        }
+      }
+
+      final bin = OneHotToBinary(oneHotList.rswizzle()).binary;
+      _output = addOutput('findFirstOne', width: bin.width);
```
Comment (mkorbel1) [inline lib/src/find.dart:74]: name of the output shouldn't say "One" since it could be 1s or 0s

File: lib/src/find.dart [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+          // If `bus[i]` is a `0` and the number of `0`'s from index 0 to `i`
+          // is `n`
+          oneHotList.add(
+              (one ? bus[i] : ~bus[i]) & paddedCountValue.eq(paddedNValue));
+        }
+      }
+
+      final bin = OneHotToBinary(oneHotList.rswizzle()).binary;
+      _output = addOutput('findNthOne', width: bin.width);
```
Comment (mkorbel1) [inline lib/src/find.dart:58]: name of output shouldn't say "One" since it could be 1s or 0s

File: lib/src/find.dart [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        if (i == 0) {
+          oneHotList.add(busCheck);
+        } else {
+          final rangeCheck =
+              one ? ~bus.getRange(0, i).or() : bus.getRange(0, i).and();
+          oneHotList.add(busCheck & rangeCheck);
+        }
+      }
+
+      final bin = OneHotToBinary(oneHotList.rswizzle()).binary;
```
Comment (mkorbel1) [inline lib/src/find.dart:73]: This code with the `OneHotToBinary` is identical between the two cases, could be pulled out to remove code duplication.

File: lib/src/find.dart [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+      _output = addOutput('findFirstOne', width: bin.width);
+      _output <= bin;
+      return;
```
Comment (mkorbel1) [inline lib/src/find.dart:76]: why `return`? i think we can delete this

File: test/count_test.dart [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+//
+// 2023 June 8
+// Author: Max Korbel <max.korbel@intel.com>
+//
+
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/src/count.dart';
+import 'package:test/test.dart';
+
+void main() {
```
Comment (mkorbel1) [inline test/count_test.dart:15]: we should test some edge cases here like all 1's and all 0's too

File: test/count_test.dart [hunk 15]
```diff
@@ -0,0 +1,27 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// find_test.dart
```
Comment (mkorbel1) [inline test/count_test.dart:4]: fix up the file header (file name, description, date, author)

File: test/find_test.dart [hunk 16]
```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// find_test.dart
+// Tests for Find
+//
+// 2023 June 8
+// Author: Max Korbel <max.korbel@intel.com>
```
Comment (mkorbel1) [inline test/find_test.dart:8]: fix author :)

File: test/find_test.dart [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+//
+
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+import 'package:rohd_hcl/src/utils.dart';
+import 'package:test/test.dart';
+
+void main() {
```
Comment (mkorbel1) [inline test/find_test.dart:16]: we should test some more edge cases:
- what if N is 0 (specified, not null)?
- what if the first one found is the 0th bit?
- what if the first one found is the last bit?
- what if all 0's?
- what if all 1's?
- what if nothing is found? do we need another output for "not found"?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]
Great progress!

File: lib/rohd_hcl.dart [hunk 18]
```diff
@@ -4,6 +4,7 @@
 export 'src/arbiter.dart';
 export 'src/exceptions.dart';
 export 'src/fifo.dart';
+export 'src/find.dart';
```
[the remark this answers is not in the document]
Comment (RPG-coder-intc) [reply lib/rohd_hcl.dart:8]: done

File: lib/src/find.dart (hunk 8 above)
Comment (mkorbel1) [reply lib/src/find.dart:40]: i thought we have a corner case when i ==0, isn't that why the if statement is here?

File: test/count_test.dart [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    final bus = Const(bin('001101'), width: 6);
+    final mod = Count(bus, countOne: false);
+    expect(mod.index.value.toInt(), 3);
+  });
+  test('count all 0s when input is all 1s', () {
+    final bus = Const(bin('11111'), width: 5);
+    final mod = Count(bus, countOne: false);
+    expect(mod.index.value.toInt(), 0);
+  });
+  test('count all 1s when input is all 0s', () {
```
Comment (mkorbel1) [inline test/count_test.dart:43]: name of the test needs to be updated

**Comment by mkorbel1** [comment]
Might want to pull from main to make sure no merge conflicts

File: test/find_test.dart (hunk 17 above)
Comment (RPG-coder-intc) [reply test/find_test.dart:16]: done

File: lib/src/one_hot.dart [hunk 20]
```diff
@@ -29,22 +29,28 @@ class BinaryToOneHot extends Module {
...
   Logic get binary => output('binary');
+  Logic get error => output('error');
 
   /// Constructs a [Module] which decodes a one-hot or thermometer-encoded
   /// number [onehot] into a 2s complement number [binary] by encoding
   /// the position of the '1'
   OneHotToBinary(Logic onehot) {
     onehot = addInput('onehot', onehot, width: onehot.width);
     addOutput('binary', width: log2Ceil(onehot.width + 1));
+    addOutput('error');
```
Comment (mkorbel1) [inline lib/src/one_hot.dart:40]: Ideally we could make the error signal optional (see `fifo.dart` for an example), then `error` would be a `Logic?` depending on whether `generateError` was true or false.  If you want to punt this to another issue/PR/someone else that's fine too

File: test/find_test.dart [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+              final bus = Const(bin('00000000'), width: 8);
+              final mod =
+                  Find(bus, countOne: false, n: Const(10, width: log2Ceil(10)));
+              expect(mod.error.value.toInt(), 1);
+            }),
+            test('if all 0s', () {
+              final bus = Const(bin('00000000'), width: 8);
+              final mod =
+                  Find(bus, countOne: false, n: Const(7, width: log2Ceil(8)));
+              expect(mod.index.value.toInt(), 0);
```
Comment (mkorbel1) [inline test/find_test.dart:93]: shouldn't the 7th `0` be in position `7` here?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]
Looking good in the right direction!

File: test/find_test.dart (hunk 21 above)
Comment (RPG-coder-intc) [reply test/find_test.dart:93]: seem n is of width `log2Ceil(8) + 1` was also missing

File: lib/src/one_hot.dart (hunk 20 above)
Comment (RPG-coder-intc) [reply lib/src/one_hot.dart:40]: done

File: lib/src/find.dart [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+      oneHotList.add(valCheck & paddedNValue.eq(paddedCountValue));
+    }
+
+    final oneHotBinary = OneHotToBinary(oneHotList.rswizzle());
+    // Upon search complete, we get the position value in binary `bin` form
+    final bin = oneHotBinary.binary;
+    addOutput('index', width: bin.width);
+    index <= bin;
+
+    addOutput('error');
```
[the remark this answers is not in the document]
Comment (RPG-coder-intc) [reply lib/src/find.dart:72]: done

File: test/find_test.dart (hunk 21 above)
Comment (RPG-coder-intc) [reply test/find_test.dart:93]: done

File: lib/src/one_hot.dart [hunk 23]
```diff
@@ -30,21 +30,43 @@ class OneHotToBinary extends Module {
...
+
     Combinational([
       Case(onehot, conditionalType: ConditionalType.unique, [
         for (var i = 0; i < onehot.width; i++)
           CaseItem(
             Const(BigInt.from(1) << i, width: onehot.width),
-            [binary < Const(i, width: binary.width)],
+            [
+              binary < Const(i, width: binary.width),
+              ...generateError
```
Comment (mkorbel1) [inline lib/src/one_hot.dart:60]: You could write this as
```
if(generateError) error < 0,
```

**Review by mkorbel1 — APPROVED** [review]
